File: integration-test/src/test/scala/ru/sberbank/blockchain/cnft/wallet/test/RestoreWallet.scala
package ru.sberbank.blockchain.cnft.wallet.test

import io.cucumber.scala.{EN, ScalaDsl}
import org.slf4j.LoggerFactory
import ru.sberbank.blockchain.cnft.common.types._
import ru.sberbank.blockchain.cnft.commons.LoggingSupport
import ru.sberbank.blockchain.cnft.model._
import ru.sberbank.blockchain.cnft.wallet.dsl._
import ru.sberbank.blockchain.cnft.wallet.spec.CNFTWalletSpec
import ru.sberbank.blockchain.cnft.wallet.store.HDPathStore
import ru.sberbank.blockchain.cnft.wallet.test.Util.{mkProfilesString, mkTokensString}
import ru.sberbank.blockchain.cnft.wallet.walletmodel.{CreateProfileInfo, WalletIssueTokenRequest}
import ru.sberbank.blockchain.cnft.wallet.{CNFT, CNFTCrypto}

import java.util.UUID
import scala.collection.mutable


/**
 * @author Alexey Polubelov
 */
class RestoreWallet extends ScalaDsl with EN with LoggingSupport {

    private val log = LoggerFactory.getLogger(classOf[RestoreWallet])

    //

    //    After { _: Scenario =>
    //        log.info("Stopping wallets ...")
    //        scenarioContext.wallets.values.foreach(_.stopListenBlocks())
    //        log.info("All wallets stopped.")
    //    }

    Given("""Create wallet for {string}""") {
        clientName: String =>
            log.info(s"Creating new wallet for $clientName")
            val CNFTGateUrl = sys.env.getOrElse("CNFT_GATE_URL", "http://localhost:8981")
            log.info(s"Using CNFT gateway: $CNFTGateUrl")
            val chain = CNFT.connect(CNFTGateUrl)

            val adminWallet = chain.connectWallet("http://localhost:8983")

            {
                for {
                    // register wallet through platform admin wallet
                    crypto <- CNFTCrypto
                        .newContext(
                            CNFTCrypto.hdSignatureOperations(HDPathStore.inMemory()),
                            CNFTCrypto.bouncyCastleEncryption(),
                            CNFTCrypto.bouncyCastleAccessOperations(),
                            CNFTCrypto.hdSignatureOperations(HDPathStore.inMemory()),
                            CNFTCrypto.hash(),
                            CNFTCrypto.secureRandomGenerator
                        )
                        .create()

                    member <- crypto.memberInformation()
                    _ <- chain.getMember(member.id) match {
                        case Left(_) =>
                            logger.info("Member does not registered yet")
                            adminWallet.registerMember(member)
                                .map { tx =>
                                    logger.info(s"Registered self [${member.id}] in TX: [${tx.blockNumber} : ${tx.txId}]")
                                }

                        case Right(_) =>
                            logger.info("Member already registered")
                            Result(())
                    }
                    wallet <- chain.newWallet(crypto)

                    //
                    // TODO: await registered?
                    //                _ = {
                    //                    log.info("Awaiting admin to register wallet... ")
                    //                    registered.get()
                    //                    log.info("Wallet registered")
                    //                }

                    walletBytes <- crypto.exportData()
                } yield {
                    scenarioContext.walletInfo.put(clientName, walletBytes)
                    //                scenarioContext.waiters.put(clientName, waiters)
                    scenarioContext.wallets.put(clientName, wallet)
                    scenarioContext.wallets.put("Admin", adminWallet)
                    //                    scenarioContext.adminWallet
                    log.info(s"Created new client $clientName")
                }
            }.orFail("Failed to create wallet")
    }

    When("""Admin creates new address""") {
        () =>
            log.info("creating address via remote wallet")
            for {
                adminWallet <- scenarioContext.wallets.get("Admin").toRight("no admin wallet found")
                _ <- adminWallet.createAddress
            } yield log.info("success")

    }

    When("""{string} registered new token type""") {
        (clientName: String) =>
            val tokenTypeName = "T1"
            log.info(s"Registering token type $tokenTypeName for $clientName ...")
            (for {
                wallet <- scenarioContext.wallets.get(clientName).toRight(s"Unknown client $clientName")
                typeId <- wallet.createId
                _ <- wallet
                    .registerTokenType(typeId,
                        TokenTypeMeta(
                            description = Collection(DescriptionField("Value", FieldType.Text, tokenTypeName)),
                            fields = Collection(FieldMeta("Value", FieldType.Text))
                        ),
                        DNA(
                            emission =
                                Collection(
                                    Gene(
                                        id = GeneID.EmissionControlledByIssuer,
                                        parameters = Collection.empty
                                    )
                                ),
                            Collection.empty,
                            Collection.empty,
                            Collection.empty),
                        Collection.empty,
                        Collection.empty
                    )
            } yield {
                scenarioContext.tokenTypeIdByName.put(tokenTypeName, typeId)
                log.info(s"Registered token type $tokenTypeName for $clientName ($typeId).")
            }).orFail(s"Failed to register token type $tokenTypeName")
    }


    When("""{string} issue new token""") {
        (issuerName: String) =>
            val tokenTypeName = "T1"
            log.info(s"Issuing token of type $tokenTypeName for $issuerName ...")
            (for {
                issuerWallet <- scenarioContext.wallets.get(issuerName).toRight(s"Unknown client $issuerName")
                //waiters <- scenarioContext.waiters.get(issuerName).toRight(s"No waiters for $issuerName")
                typeId <- scenarioContext.tokenTypeIdByName.get(tokenTypeName).toRight(s"Unknown token type $tokenTypeName")
                toId <- issuerWallet.getIdentity
                owner <- issuerWallet.createSingleOwnerAddress
                tokenId <- issuerWallet.createTokenId(typeId)
                await = Util.awaitWalletEvents(issuerWallet) { case (_, events) =>
                    if (events.owner.tokensReceived.nonEmpty)
                        Some(events.owner.tokensReceived)
                    else
                        None
                }
                _ <- issuerWallet
                    .issue(
                        Collection(
                            WalletIssueTokenRequest(
                                tokenId,
                                owner,
                                TokenContent(
                                    Collection("100")
                                ),
                                toId
                            )
                        )
                    )
            } yield {
                log.info("Awaiting tokens to arrive ...")
                val added = await.get
                log.info(s"Got tokens: ${added.mkString(", ")}")

                scenarioContext.tokenIdByName.put("T1", tokenId)
                log.info(s"Issued token of type $tokenTypeName for $issuerName ($tokenId).")
                ()
            }).orFail("Error when issuing new token")
    }

    When("""{string} create new profile""") {
        (issuerName: String) =>
            log.info(s"Creating profile for $issuerName ...")
            (for {
                wallet <- scenarioContext
                    .wallets
                    .get(issuerName)
                    .toRight(s"Unknown client $issuerName")
                profileName = UUID.randomUUID().toString
                _ <- wallet.createProfile(
                    CreateProfileInfo(
                        profileName,
                        "test_description",
                        "test_avatar",
                        "test_background"
                    )
                )
            } yield {
                log.info(s"Profile was successfully created for $issuerName")
            }).orFail(s"Failed to create profile for $issuerName")
    }

    When("""Restore wallet for client {string}""") { (clientName: String) =>

        log.info(s"Creating new wallet for $clientName")
        val CNFTGateUrl = sys.env.getOrElse("CNFT_GATE_URL", "http://localhost:8981")
        log.info(s"Using CNFT gateway: $CNFTGateUrl")

        {
            for {
                walletInfo <- scenarioContext.walletInfo.get(clientName).toRight(s"Unknown client $clientName")
                context =
                    CNFTCrypto
                        .newContext(
                            CNFTCrypto.hdSignatureOperations(HDPathStore.inMemory()),
                            CNFTCrypto.bouncyCastleEncryption(),
                            CNFTCrypto.bouncyCastleAccessOperations(),
                            CNFTCrypto.hdSignatureOperations(HDPathStore.inMemory()),
                            CNFTCrypto.hash(),
                            CNFTCrypto.secureRandomGenerator
                        )
                crypto <- context.importFrom(walletInfo)

                chain = CNFT.connect(CNFTGateUrl)
                wallet <- chain.newWallet(crypto)
            } yield {
                scenarioContext.wallets.put(clientName, wallet)
                log.info(s"Restore wallet new client $clientName")
            }
        }.orFail(s"Error while restoring wallet for client $clientName")
    }

    When("""Check token list in wallet for client {string}""") {
        (clientName: String) =>
            log.info(s"Checks tokens for $clientName...")
            (for {
                wallet <- scenarioContext.wallets.get(clientName).toRight(s"Unknown client $clientName")
                _ = Thread.sleep(5000)
                tokens <- wallet.listTokens
            } yield log.info(mkTokensString(tokens))).orFail(s"Failed to check tokens for $clientName")
    }

    When("""Check {int} tokens in wallet for client {string}""") {
        (expectedCount: Int, clientName: String) =>
            log.info(s"Checks tokens for $clientName...")
            (for {
                wallet <- scenarioContext.wallets.get(clientName).toRight(s"Unknown client $clientName")
                tokens <- wallet.listTokens
            } yield {
                log.info(mkTokensString(tokens))
                assert(tokens.length == expectedCount)
            }).orFail(s"Failed to check tokens for $clientName")
    }

    When("""Check {int} profiles in wallet for client {string}""") {
        (expectedCount: Int, clientName: String) =>
            log.info(s"Checks profiles for $clientName...")
            (for {
                wallet <- scenarioContext.wallets.get(clientName).toRight(s"Unknown client $clientName")
                profiles <- wallet.listProfiles
            } yield {
                log.info(mkProfilesString(profiles))
                assert(profiles.length == expectedCount)
            }).orFail(s"Failed to check tokens for $clientName")
    }

    private val _context = new ThreadLocal[ScenarioContext]() {
        override def initialValue(): ScenarioContext = ScenarioContext()
    }

    private def scenarioContext: ScenarioContext = _context.get()

    case class ScenarioContext(
        wallets: mutable.Map[String, CNFTWalletSpec[Result]] = mutable.HashMap.empty[String, CNFTWalletSpec[Result]],
        tokenTypeIdByName: mutable.Map[String, String] = mutable.HashMap.empty[String, String],
        tokenIdByName: mutable.Map[String, String] = mutable.HashMap.empty[String, String],
        walletInfo: mutable.Map[String, Bytes] = mutable.HashMap.empty[String, Bytes],
    )

}

File: integration-test/src/test/js/restore-wallet.js
const { profilesContextMixin, clientCreatesProfile, walletContextMixin } = require('./cnft-common-steps');
const { log } = require('./cnft-util');
const { Given, When, After, setDefaultTimeout } = require('@cucumber/cucumber');
const { fail, strictEqual } = require('assert');
const {
    CNFT,
    CNFTCrypto,
    CreateProfileInfo,
    DNA,
    DescriptionField,
    FieldMeta,
    FieldType,
    Gene,
    GeneID,
    HDPathStore,
    TokenContent,
    TokenTypeMeta,
    WalletIssueTokenRequest,
} = require('wallet-lib');

setDefaultTimeout(10 * 1000);

global.window = {}
window.localStorage = global.localStorage

const elliptic = require('elliptic');
const EC = elliptic.ec
const jsec = new EC('secp256k1')

const crypto = require('crypto').webcrypto;
// export WebCrypto to global scope, so it to be accessible by XXXCryptoOperations
global.crypto = crypto;

global.Buffer = global.Buffer || require('buffer').Buffer;
const XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest;
// export XMLHttpRequest to global scope, so it to be accessible by GateApi:
global.XMLHttpRequest = XMLHttpRequest;
let scenarioContext = {
    "walletSeeds": {},
    "walletIds": {},
    "walletInfo": {},
    "tokenTypeIdByName": {},
};
Object.assign(scenarioContext, walletContextMixin);
Object.assign(scenarioContext, profilesContextMixin);

Given('there is embeded postgres',
    async function () {
        fail(`Non applicable for JS test`);
    }
)

async function awaitWalletEvents(wallet, f) {
    var running = true
    var lastSeenBlock = await wallet.chain.getLatestBlockNumber
    log(`Starting to wait event, block: [${lastSeenBlock}]`);
    while (running) {
        let height = await wallet.chain.getLatestBlockNumber
        while (lastSeenBlock <= height) {
            let events = await wallet.events(lastSeenBlock, false)
            let result = f(wallet, events)
            log(`Processed block [${lastSeenBlock}], found: ${result}`);
            if(result)
                return events
            lastSeenBlock = (parseInt(lastSeenBlock) + 1)
        }
        await new Promise((resolve) => {
            setTimeout(resolve, 1000)
        })
    }
}

Given('Create wallet for {string}',
    async function (clientName) {
        log(`Creating new wallet for ${clientName}`)
        let CNFTGateUrl = "http://localhost:8981"

        let context =
            CNFTCrypto.newContext(
                CNFTCrypto.hdSignatureOperations(crypto, jsec, HDPathStore.inMemory()),
                CNFTCrypto.webCryptoEncryption(crypto),
                CNFTCrypto.webCryptoAccessOperations(crypto, jsec),
                CNFTCrypto.hdSignatureOperations(crypto, jsec, HDPathStore.inMemory()),
                CNFTCrypto.hash(crypto),
                CNFTCrypto.secureRandomGenerator(crypto)
            );
        let walletCrypto = await context.create();

        // save wallet bytes to context:
        let walletBytes = await walletCrypto.exportData();
        scenarioContext.walletSeeds[clientName] = walletBytes;

        //
        log("Connecting to Admin wallet ...")
        let chain = await CNFT.connect(CNFTGateUrl)
        let adminWallet = await chain.connectWallet("http://localhost:8983");
        let member = await walletCrypto.memberInformation();
        log("Registering wallet ...")
        let tx = await adminWallet.registerMember(member);
        log(`Registered wallet [${member.id}] in TX: [${tx.blockNumber} : ${tx.txId}]`)
        //
        await chain
            .newWallet(walletCrypto)
            .then(async wallet => {
                await wallet.getWalletInformation.then(async walletIdentity => {
                    scenarioContext.walletIds[clientName] = walletIdentity.id;
                    scenarioContext.wallets[clientName] = wallet;
                    scenarioContext.walletInfo[clientName] = walletIdentity;
                });
            });
    }
);

When('{string} registered new token type',
    async function (clientName) {
        let tokenTypeName = "T1"
        log(`Registering token type ${tokenTypeName} for ${clientName} ...`);
        let wallet = scenarioContext.wallets[clientName];
        if (typeof wallet === 'undefined') fail(`Unknown client ${clientName}`);
        let typeId = await wallet.createId;

        let result = await wallet.registerTokenType(
            typeId,
            new TokenTypeMeta(
                [
                    new DescriptionField("value", FieldType.Text, tokenTypeName),
                    new DescriptionField("other", FieldType.Text, "О-ГО-ГО!")
                ],
                [new FieldMeta("Value", FieldType.Text)]
            ),
            DNA(
                [
                    Gene(GeneID.EmissionControlledByIssuer, [])
                ],
                [
                    Gene(GeneID.RequireOwnerSignatureForTransfer, [])
                ],
                [
                    Gene(GeneID.RequireOwnerSignatureForBurn, [])
                ],
                [
                    Gene(GeneID.Fungible, [])
                ]
            ),
            [], []
        )
        scenarioContext.tokenTypeIdByName[clientName] = typeId;
        log(`Registered token type ${tokenTypeName} for ${clientName} (${typeId}).`)
        return result;
    }
);

When('{string} issue new token', { timeout: 200000 }, async function (issuerName) {
    log(`Issuing token for ${issuerName}...`)
    let tokenTypeName = "T1"
    let clientWallet = scenarioContext.wallets[issuerName];
    if (typeof clientWallet === 'undefined') fail(`Unknown issuerName  ${issuerName}`);
    let tokenValue = "100";
    let owner = await clientWallet.createSingleOwnerAddress;
    let tokenType = scenarioContext.tokenTypeIdByName[issuerName];
    let tokenId = await clientWallet.createTokenId(tokenType)
    let myId = await clientWallet.getIdentity;
    let requests = [
        new WalletIssueTokenRequest(
            tokenId,
            owner,
            new TokenContent(
                [tokenValue]
            ),
            myId
        )
    ]
    if (tokenId !== undefined) {
        let waitIssuedTokens = awaitWalletEvents(clientWallet, (wallet, events) => {
            if (
                events.owner.tokensReceived.length > 0 &&
                events.owner.tokensReceived.includes(tokenId)
            ) {
                log(`Tokens Received [${events.owner.tokensReceived.toString()}]`);
                return true;
            }
            return false;
        })

        clientWallet
            .issue(
                requests
            ).catch(function (err) {
                log(err)
        })
        scenarioContext.tokenTypeIdByName["T1"] = tokenId;
        await waitIssuedTokens;
        log(`Issued token of type ${tokenTypeName} for ${issuerName}`);
    }
});

When('{string} create new profile', async function (issuerName) {
    var random_num = new Uint8Array(10); // 2048 = number length in bits
    crypto.getRandomValues(random_num);
    let profileName = random_num.toString(36).substring(7);
    const created = clientCreatesProfile(
        scenarioContext,
        issuerName,
        CreateProfileInfo(
            profileName,
            'test_description',
            'test_avatar',
            'test_background'
        )
    );
    await created;
});

When('Restore wallet for client {string}',
    async function (clientName) {
        log(`Creating new wallet for ${clientName}`)
        let CNFTGateUrl = "http://localhost:8981"

        let context =
            CNFTCrypto.newContext(
                CNFTCrypto.hdSignatureOperations(crypto, jsec, HDPathStore.inMemory()),
                CNFTCrypto.webCryptoEncryption(crypto),
                CNFTCrypto.webCryptoAccessOperations(crypto, jsec),
                CNFTCrypto.hdSignatureOperations(crypto, jsec, HDPathStore.inMemory()),
                CNFTCrypto.hash(crypto),
                CNFTCrypto.secureRandomGenerator(crypto)
            );
        let walletBytes = scenarioContext.walletSeeds[clientName];
        let walletCrypto = await context.importFrom(walletBytes);

        //
        let chain = await CNFT.connect(CNFTGateUrl)
        let wallet = await chain.newWallet(walletCrypto)

        const walletInformation = await wallet.getWalletInformation;

        scenarioContext.walletIds[clientName] = walletInformation.id;
        scenarioContext.wallets[clientName] = wallet;
        scenarioContext.walletInfo[clientName] = walletInformation;

        log(`Restored wallet for client ${clientName}`);
    }
);

When('Check token list in wallet for client {string}', async function (clientName) {
    log(`Checks tokens for ${clientName} ...`)

    function sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    await sleep(2000)
    let wallet = scenarioContext.wallets[clientName];
    let tokens = await wallet.listTokens
    log(tokens)
});

When('Check {int} tokens in wallet for client {string}',
    async function (expectedCount, clientName) {
        log(`Checks tokens for ${clientName} ...`);
        let wallet = scenarioContext.wallets[clientName];
        let tokens = await wallet.listTokens;
        strictEqual(tokens.length, expectedCount);
    });

When('Check {int} profiles in wallet for client {string}',
    async function (expectedCount, clientName) {
        log(`Checks profiles for ${clientName} ...`);
        let wallet = scenarioContext.wallets[clientName];
        let profiles = await wallet.listProfiles;
        strictEqual(profiles.length, expectedCount);
    });

When('{string} checks operations list',
    async function (clientName) {
        let wallet = scenarioContext.wallets[clientName];
        let walletIdentity = await wallet.getIdentity;
        log('');
        log(`Operations for ${walletIdentity}`);

        log('_______')
        let operations = await wallet.listOperations;
        operations.forEach(op => {
            log(`id: ${op.operationId}`);
            log(`history:`);
            op.history.forEach(h => {
                log(`opstate: ${h.state}, timestamp: ${h.timestamp}, txId: ${h.txId}`);
            });
        });
    });


/*After(
    function () {
        return new Promise((resolve, _reject) => {
            log(`Stopping wallets ...`);
            for (const index in scenarioContext.wallets) {
                log(`Stopping wallet ${index} ...`);
            }
            log(`Wallets stopped.`);
            return resolve();
        });
    }
);*/
